safehttp/form.go

```diff
@@ -0,0 +1,249 @@
+	}
+	return vals[0]
+}
+
+// Float64 checks whether key paramKey maps to any query or form parameter
+// values. In case it does, it will try to convert the first value to a float
+// and return it. If there are no values associated with paramKey, it will
+// return the default value. If the first value is not a float, it will return
+// the default value and set the Form error field.
+func (f *Form) Float64(paramKey string, defaultValue float64) float64 {
+	if f.err != nil {
+		return defaultValue
+	}
+	vals, ok := f.values[paramKey]
+	if !ok {
+		return defaultValue
+	}
+	paramVal, err := strconv.ParseFloat(vals[0], 64)
+	if err != nil {
+		f.err = err
+		return defaultValue
+	}
+	return paramVal
+}
+
+// Bool checks whether key paramKey maps to any query  or form parameter
+// values. In case it does, it will try to convert the first value to a boolean
+// and return it. If there are no values associated with paramKey, it will
+// return the default value. If the first value is not a boolean, it will return
+// the default value and set the Form error field.
+func (f *Form) Bool(paramKey string, defaultValue bool) bool {
+	if f.err != nil {
+		return defaultValue
+	}
+	vals, ok := f.values[paramKey]
+	if !ok {
+		return defaultValue
+	}
+	switch vals[0] {
+	case "true":
+		return true
+	case "false":
+		return false
+	default:
+		f.err = fmt.Errorf("values of form parameter %q not a boolean", paramKey)
+	}
+	return false
+}
+
+func clearSlice(slicePtr interface{}) error {
+	switch vs := slicePtr.(type) {
+	case *[]string:
+		*vs = nil
+	case *[]int:
+		*vs = nil
+	case *[]float64:
+		*vs = nil
+	case *[]uint64:
+		*vs = nil
+	case *[]bool:
+		*vs = nil
+	default:
+		return fmt.Errorf("type not supported in Slice call: %T", vs)
+	}
+	return nil
+}
+
+// Slice checks whether key paramKey maps to any query or form parameters. If it
+// does, it will try to convert them to the type of slice elements slicePtr
+// points to. If there are no values associated with paramKey, it will clear the
+// slice. If type conversion fails at any point, the Form error field will be
+// set and the slice will be cleared.
+func (f *Form) Slice(slicePtr interface{}, paramKey string) {
+	if f.err != nil {
+		f.err = clearSlice(slicePtr)
+		return
+	}
+	mapVals, ok := f.values[paramKey]
+	if !ok {
+		f.err = clearSlice(slicePtr)
+		return
+	}
+	switch values := slicePtr.(type) {
+	case *[]string:
+		res := make([]string, 0, len(mapVals))
+		for _, x := range mapVals {
+			res = append(res, x)
+		}
+		*values = res
+	case *[]int:
+		res := make([]int, 0, len(mapVals))
+		for _, x := range mapVals {
+			x, err := strconv.Atoi(x)
+			if err != nil {
+				f.err = err
+				*values = nil
+				return
+			}
+			res = append(res, x)
+			*values = res
+		}
+	case *[]uint64:
+		res := make([]uint64, 0, len(mapVals))
+		for _, x := range mapVals {
+			x, err := strconv.ParseUint(x, 10, 0)
+			if err != nil {
+				f.err = err
+				*values = nil
+				slicePtr = values
```

Comment:
Should this row be removed? You are already setting `slicePtr` to nil by setting `*values = nil` on the row above

Reply:
good spot ^^